File: src/main/java/core/scm/specialforms/If.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.exceptions.IllegalSyntaxException;
import core.scm.Thunk;
import core.utils.Utils;

import java.util.List;

/* Syntax:
 * (if <test> <consequent> <alternate>)
 * (if <test> <consequent>)
 */
public enum If implements ISpecialForm {
  IF;

  @Override
  public Object eval(List<Object> expression, Environment env, Evaluator evaluator) {
    int size = expression.size();
    if (size != 4) {
      throw IllegalSyntaxException.of(toString(), expression, String.format("has %s parts after keyword", size - 1));
    }
    return Utils.INSTANCE.toBoolean(evaluator.eval(expression.get(1), env)) ?
           new Thunk(expression.get(2), env) :
           new Thunk(expression.get(3), env);
  }

  @Override
  public String toString() {
    return "if";
  }
}

File: src/main/java/core/scm/specialforms/If.kt
package core.scm.specialforms

import core.Evaluator
import core.exceptions.IllegalSyntaxException
import core.scm.Thunk
import core.utils.Utils
import core.Writer

/**
 * Syntax:
 * (if <test> <consequent> <alternate>)
 * (if <test> <consequent>)
 */
object If : SpecialForm("if") {

    override fun eval(form: List<Any?>, evaluator: Evaluator): Any = when {
        form.size != 4 -> throw IllegalSyntaxException(toString(), Writer.write(form), "has ${form.size - 1} parts after keyword")
        else -> when {
            Utils.toBoolean(evaluator.eval(form[1])) -> Thunk(form[2], evaluator.env)
            else -> Thunk(form[3], evaluator.env)
        }
    }
}
